### PR TITLE
Handle empty arrays in a variety of functions

### PR DESCRIPTION
**Pull request by syurkevi**
Allows indexing with empty arrays . Also handles weird cases such as `af::det( [ ] ) = 1` 

- Fixes https://github.com/arrayfire/arrayfire/issues/799

### CODE REVIEW

File: src/api/c/assign.cpp [hunk 1]
```diff
@@ -233,6 +232,11 @@ af_err af_assign_gen(af_array *out,
         af_dtype lhsType= lInfo.getType();
         af_dtype rhsType= rInfo.getType();
 
+        if(rhsDims.ndims() == 0) {
+            dim_t my_dims[] = { 0, 0, 0, 0 };
+            return af_create_handle(out, AF_MAX_DIMS, my_dims, rhsType);
```
Comment (pavanky) [inline src/api/c/assign.cpp]: This is incorrect. We should check if both rhs and index are empty and then return a retained version of `lhs`.

Comment (pavanky) [inline src/api/c/assign.cpp]: If `lhs` is empty, then return the way you did here.

File: src/api/c/convolve.cpp [hunk 2]
```diff
@@ -72,6 +72,11 @@ af_err convolve(af_array *out, const af_array signal, const af_array filter)
         dim4 sdims = sInfo.dims();
         dim4 fdims = fInfo.dims();
 
+        if(fdims.ndims() == 0 || sdims.ndims() ==  0) {
+            af_retain_array(out, signal);
+            return AF_SUCCESS;
```
Comment (pavanky) [inline src/api/c/convolve.cpp]: return af_retain_array(out, signal);

File: src/api/c/det.cpp [hunk 3]
```diff
@@ -78,19 +78,30 @@ af_err af_det(double *real_val, double *imag_val, const af_array in)
 
         cfloat cfval;
         cdouble cdval;
+        bool empty = i_info.dims()[0] == 0;
 
         switch(type) {
-        case f32: *real_val = det<float  >(in);  break;
-        case f64: *real_val = det<double >(in);  break;
+        case f32: *real_val = (empty)? 1.f : det<float  >(in);  break;
```
Comment (pavanky) [inline src/api/c/det.cpp]: Perhaps do the checks in the wrapper function. That way you can return scalar<T>(1.0); once and not worry about multiple if/elese statements?

File: src/api/c/diff.cpp [hunk 4]
```diff
@@ -40,6 +40,12 @@ af_err af_diff1(af_array *out, const af_array in, const int dim)
         af_dtype type = info.getType();
 
         af::dim4 in_dims = info.dims();
+        if(in_dims[dim] < 2) {
+            dim_t my_dims[] = {0, 0, 0, 0};
+            AF_CHECK(af_create_handle(out, AF_MAX_DIMS, my_dims, type));
```
Comment (pavanky) [inline src/api/c/diff.cpp]: Perhaps this should be retain rather than create_handle ?

File: src/api/c/diff.cpp [hunk 5]
```diff
@@ -77,6 +83,11 @@ af_err af_diff2(af_array *out, const af_array in, const int dim)
         af_dtype type = info.getType();
 
         af::dim4 in_dims = info.dims();
+        if(in_dims[dim] < 3) {
+            dim_t my_dims[] = {0, 0, 0, 0};
+            AF_CHECK(af_create_handle(out, AF_MAX_DIMS, my_dims, type));
```
Comment (pavanky) [inline src/api/c/diff.cpp]: Same goes here

File: src/api/c/fft.cpp [hunk 6]
```diff
@@ -33,6 +33,11 @@ static af_err fft(af_array *out, const af_array in, const double norm_factor, co
         af_dtype type  = info.getType();
         af::dim4 dims  = info.dims();
 
+        if(dims.ndims() == 0) {
+            AF_CHECK(af_retain_array(out, in));
+            return AF_SUCCESS;
```
Comment (pavanky) [inline src/api/c/fft.cpp]: return af_retain_array. No reason for AF_CHECK here.

File: src/api/c/fft.cpp [hunk 7]
```diff
@@ -215,6 +227,10 @@ static af_err fft_c2r(af_array *out, const af_array in, const double norm_factor
         af_dtype type  = info.getType();
         af::dim4 idims  = info.dims();
 
+        if(idims.ndims() == 0) {
+            AF_CHECK(af_retain_array(out, in));
+            return AF_SUCCESS;
+        }
```
Comment (pavanky) [inline src/api/c/fft.cpp]: goes for all of them in this file.

File: src/api/c/qr.cpp [hunk 8]
```diff
@@ -41,6 +42,7 @@ static inline af_array qr_inplace(af_array in)
 
 af_err af_qr(af_array *q, af_array *r, af_array *tau, const af_array in)
 {
+    std::cout<<"Isqrentrrr"<<std::endl;
```
Comment (pavanky) [inline src/api/c/qr.cpp]: Don't forget to remove this.

File: src/api/c/qr.cpp [hunk 9]
```diff
@@ -50,6 +52,15 @@ af_err af_qr(af_array *q, af_array *r, af_array *tau, const af_array in)
 
         af_dtype type = i_info.getType();
 
+        std::cout<< " amihere"<<(i_info.ndims() == 0)<<std::endl;
```
Comment (pavanky) [inline src/api/c/qr.cpp]: Don't forget to remove this.

File: src/api/cpp/histogram.cpp [hunk 10]
```diff
@@ -26,6 +26,9 @@ array histogram(const array &in, const unsigned nbins, const double minval, cons
 array histogram(const array &in, const unsigned nbins)
 {
     af_array out = 0;
+    if(in.numdims() == 0) {
+        return in;
+    }
```
Comment (pavanky) [inline src/api/cpp/histogram.cpp]: Why is this here and not in the C API ?

**Comment by shehzan10** [comment]
May I suggest tests for all of these?

File: src/api/cpp/histogram.cpp (hunk 10 above)
Comment (syurkevi) [inline src/api/cpp/histogram.cpp]: Because the min<T> and max<T> will throw errors with | in | == 0. Checking here let's us return an empty array. 
I did add checks for the pure C API.

**Comment by shehzan10** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Comment by shehzan10** [comment]
build arrayfire linux ci
build arrayfire osx ci
build arrayfire windows ci

**Comment by shehzan10** [comment]
build arrayfire windows ci
